### PR TITLE
ashift: a module for automatic perspective correction

### PR DESCRIPTION
**Pull request by upegelow**
This PR aims to develop a module for automatic perspective corrections. Although the clipping module already has a feature for manual adjustments it tends to be tedious work if only scattered vertical or horizontal features are found in the image.

This module is inspired by the program ShiftN (http://www.shiftn.de) developed by Marcus Hebel.
The idea is to operate a line detection algorithm searching for vertical and/or horizontal lines and then fit the perspective correction parameters for an optimal effect.

Right now the basic manual functionality is working, line detection and parameter fitting still to be added.

### CODE REVIEW

File: src/iop/ashift.c [hunk 1]
```diff
@@ -0,0 +1,938 @@
...
+    for(int i = 0; i < 3; i++)
+    {
+      float x = 0.0f;
+      for(int j = 0; j < 3; j++) x += m1[3 * k + j] * m2[3 * j + i];
+      dst[3 * k + i] = x;
+    }
+  }
+}
+
+void _print_roi(const dt_iop_roi_t *roi, const char *label)
```
Comment (LebedevRI) [inline src/iop/ashift.c]: static

Comment (upegelow) [inline src/iop/ashift.c]: mat3mul() etc. should go static and _print_roi() should finally go away when development is finished.

**Comment by LebedevRI** [comment]
For history, i want to quote a few things:
http://article.gmane.org/gmane.comp.graphics.lensfun/384
http://wilson.bronger.org/lensfun/perspective-correction.html

In future, we would like to move all the functionality that requires interpolation (that is: keystone correction and rotation) from c&r into lens iop.

**Comment by upegelow** [comment]
Good initiative to rework c&r as it tends to be overloaded or virtually impossible to maintain (had my own experiences with that...)

Main advantage of ashift over all manual perspective corrections would be ease of use in case of not-so-clear image features (either blurry and difficult to accurately spot or too many too small ones).

**Comment by upegelow** [comment]
The module is mostly ready now. A few quirks are remaining but they shouldn't be show stoppers. Needs some testing by a broader audience so I suggest to merge soon.

**Comment by LebedevRI** [comment]
`src/iop/ashift_lsd.c` and `src/iop/ashift_nmsimplex.c` - are there any changes in them from the 'downloaded' original versions?
I would probably recommend to put them into `src/external/`

**Comment by upegelow** [comment]
Both are changed versus the original ones. The changes in LSD are minor, the ones in nmsimplex are quite significant. Changes vs. original are documented in the headers.

**Comment by LebedevRI** [comment]
github really cripples ability to comment on such big pr's..
modify_roi_out() and modify_roi_in(): you can easily vectorize those, see https://github.com/darktable-org/darktable/blob/50b7d19/src/iop/lens.c#L889
+ add `collapse(2)`

**Comment by upegelow** [comment]
In this case overhead of going multi-threaded is probably slowing down things. In contrast to lens.c this module will only make an affine transformation of the image. So the rectangular image will transfer into a tetragon with straight lines between the corners. Therefore we only need to visit those four corners to get the roi's. The for loops are just a lazy form of expressing that :) I could as well have stated the four checks individually.

**Comment by TurboGit** [comment]
Works great! Just a minor issue. The module is too large and does not fit into the right panel. I'll add a link with a picture asap...

**Comment by TurboGit** [comment]
Here it is: https://drive.google.com/file/d/0B_w_CWtzkwW0S3c0OHp3blptSFU/view?usp=sharing

**Comment by upegelow** [comment]
Thanks for reporting that. I am not very familiar with the grid packer which I needed to use it here for the first time in order to get a halfway structured layout. I decided to force all columns of the grid to the same size. Obviously the larger size of the text label now determines the total size and leads to the observed issue. Might get even worse for many translated strings. I have to see how to solve this differently. Ideally I would give the text label the space that it needs (plus some extra) and let the three other columns use all the remaining space in an equal distribution.

**Comment by LebedevRI** [comment]
> Therefore we only need to visit those four corners to get the roi's.
Oh, right :)
In all other iops similar problems were solved using `get_corner()` and `adjust_aabb()`, i did not see those functions called and did not notice that it was only checking the 4 points.
Definitely not worth parallelizing it :)

BTW: *IIRC* there are some other places in dt that might be better with that "flip detection using backtransform", namely gradient mask, graduated ND - they are kind-of wrong if flip != off.

**Comment by TurboGit** [comment]
Looks very good now! The GUI is clean and it works fine. I would merge it soon for a wider testing audience.

**Comment by LebedevRI** [comment]
```
[ 49%] Building C object src/iop/CMakeFiles/ashift.dir/introspection_ashift.c.o
In file included from /home/lebedevri/darktable/build/src/iop/introspection_ashift.c:24:
/home/lebedevri/darktable/src/iop/ashift.c:997:47: error: duplicate 'const' declaration specifier [-Werror,-Wduplicate-decl-specifier]
static void ransac(const dt_iop_ashift_line_t const* lines, int *index_set, int *inout_set,
                                              ^
1 error generated.
src/iop/CMakeFiles/ashift.dir/build.make:67: recipe for target 'src/iop/CMakeFiles/ashift.dir/introspection_ashift.c.o' failed
make[2]: *** [src/iop/CMakeFiles/ashift.dir/introspection_ashift.c.o] Error 1
make[2]: Target 'src/iop/CMakeFiles/ashift.dir/build' not remade because of errors.
CMakeFiles/Makefile2:3329: recipe for target 'src/iop/CMakeFiles/ashift.dir/all' failed
make[1]: *** [src/iop/CMakeFiles/ashift.dir/all] Error 2
```

**Comment by LebedevRI** [comment]
Rotation slider is rotating clockwise for positive values, and counterclockwise for negative...
I do see that it is not labelled as degrees, so i understand that it is kind-of not a bug, but still seems wrong.
Also, because we have other rotation sliders, which IIRC correctly rotate (clockwise for negative values), it will be confusing.

**Comment by LebedevRI** [comment]
So i have tried it, overall looks good, looking forward to it.
About those 0.5 hard limits for shift - are they really hardcoded deep down in code, or can they be easily changed to something bigger? I would prefer for them to be bigger.

**Comment by upegelow** [comment]
> About those 0.5 hard limits for shift - are they really hardcoded deep down in code, or can they be easily changed to something bigger? I would prefer for them to be bigger.

I orignally had a value of 1.0 here but then I found that all cases where you still can extract a reasonably sized rectangular image have a  lower shift value. So let's see what users report.  We can at any time increase this value if needed.

Concerning the rotation angle. I find the mathematically correct polarity a bit counter intuitive - moving the slider right turns the image "left". But I understand that we should have rotation sliders behave the same in darktable modules. So I will change that in ashift.

**Comment by LebedevRI** [comment]
>>About those 0.5 hard limits for shift - are they really hardcoded deep down in code, or can they be easily changed to something bigger? I would prefer for them to be bigger.

>I orignally had a value of 1.0 here but then I found that all cases where you still can extract a reasonably sized rectangular image have a lower shift value. So let's see what users report. We can at any time increase this value if needed.

Yes, that is exactly why bauhaus supports soft limits, see exposure iop, black level:
* [1](https://github.com/darktable-org/darktable/blob/master/src/iop/exposure.c#L816)
* [2](https://github.com/darktable-org/darktable/blob/master/src/iop/exposure.c#L450)

**Pull request merged**